🛝 Fun Times at the Playground

brettimus opened the pull request:
Creates a `playground` folder that has a SPA that looks suspiciously similar to Studio.

Current sprint work is in [Linear](https://linear.app/fiberplane/project/playground-ui-59e2f84f2ed5/issues).

It's using the designs in Figma

- [Designs](https://www.figma.com/design/RwMIgubdA8l2AyEZvPcTL9/FPX-documentation?node-id=30-2117&p=f&m=dev) 
- [Design System](https://www.figma.com/design/yI3mYPIbeupb5hdpHvLiW2/Fpx-Design-System?node-id=96-5293&p=f&m=dev)


<details open>
<summary>
<h2>Running the playground</h2>
</summary>

Easy. It uses port `6660` 👹 

- `cd playground`
- `pnpm dev`

_**Need routes?**_

- Run a fiberplane-enabled api on `localhost:8787`
- You can use this one if you want: https://github.com/brettimus/mock-music-openapi-api

<details>
<summary>
<h2>Publishing the embedded playground</h2>
</summary>

The `@fiberplane/hono` library is published with the playground assets. We serve those over the jsdelivr cdn.

To cut a new update:

First, update `packages/fiberplane-hono/package.json` to the new version.

Finally, from project root:

```sh
# This will copy the dist files from `playground` into the published package as well.
pnpm build:fp-hono
cd packages/fiberplane-hono
pnpm publish
```

Then, to integrate into an app:

```typescript
// If you have the api spec as json already, import it into your file and pass the raw, parsed spec as the `spec` property:

app.use("/fp/*", createMiddleware({
  apiKey: "12345", // This can also be picked up from the env: `FIBERPLANE_API_KEY`
  openapi: {
    url: "/doc",
  }
}));

// alternatively, you can generate the spec on the fly and pass it to the spec param (NOT RECOMMENDED)
const stringifiedSpec = JSON.stringify(app.getOpenAPIDocument({
  openapi: "3.0.0",
  info: {
    title: "My API",
    version: "0.0.1",
    description: "API documentation for My API",
  },
}));

app.use(
  "/fp/*",
  createMiddleware({
    openapi: {
      content: stringifiedSpec,
    }
  }),
); 
```
</details>

## Theming 

Needs more thought.

**See `playground/src/index.css`**

I had gemini try to convert all the tokens from figma into css variables, it may have gotten a few wrong.

I haven't put much effort into naming semantic tokens we can ultimately use in the UI alongside the shadcn ones. 

Contrast is quite low for `muted` colors right now, that likely needs some tweaking.

Using orange for the accent color can get quite aggressive sometimes. We will likely want some overrides on the default button component that comes with shadcn in order to adapt it for our new designs.

brettimus committed:
Remove nodejs support

brettimus committed:
Pass through the spec

brettimus committed:
Remove console.logs

brettimus committed:
Fix favicon

brettimus committed:
Make the docs tab work better

brettimus committed:
Update build

brettimus committed:
Add pragmas to make fp embed work inside our fp api

brettimus committed:
Publish playground with embed

brettimus committed:
Remove prepublish because it destroys things we need

brettimus committed:
Merge branch 'main' into playground

brettimus committed:
Downgrade react because of version mismatch between @fiberplane/hooks and playground

brettimus committed:
Change mock server default host to fix CORS issues

brettimus committed:
Add styles for the StatusCode component

brettimus committed:
Tweak some styles that were a little aggressive given our accent color

stephlow committed:
Merge remote-tracking branch 'origin/main' into embedded-playground-api

brettimus committed:
Add theme switching

brettimus committed:
Persist settings to local storage and fix bg-emphasis-dark

brettimus committed:
Update playground dist

brettimus committed:
Add hono as a peer dep?

brettimus committed:
Tweak styles on the json viewer

brettimus committed:
Update json editor styles

brettimus committed:
Bump embedded p.json

stephlow committed:
WIP

stephlow committed:
Merge remote-tracking branch 'origin/main' into embedded-playground-api

brettimus committed:
Check in playground dist

brettimus committed:
Update lilo-worker

brettimus committed:
Add origin of request as toplevel server

brettimus committed:
Update embedded for lilo

brettimus committed:
Update

stephlow committed:
Service implementation scaffolding

brettimus committed:
Simplify lilo fp embed

brettimus committed:
Ignore playground dist again

brettimus committed:
Default cdn url for embedded

brettimus committed:
Try something

brettimus committed:
Remove fp-status* classes and use semantic token

brettimus committed:
Simplify colors in use for http methods

brettimus committed:
Format

brettimus committed:
Small tweak

brettimus committed:
Add geist font

brettimus committed:
Change default tooltip styles to use white bg instead of primary

brettimus committed:
Faker first pass and json text editing style updates

brettimus committed:
Add tightknit api spec for testing

brettimus committed:
Update version p.json for embedded

brettimus committed:
Test 0.0.15

brettimus committed:
Update embedded in lilo

stephlow committed:
Services client and temp embedded api routes

stephlow committed:
Fix base service exports

brettimus committed:
Fix sidepaneltrigger tooltip styles

brettimus committed:
Style tweaks: Sidebar tooltip, Faker button, tab font weight

brettimus committed:
Add keyboard shortcut to faker

brettimus committed:
Fix dark mode json editing styles

brettimus committed:
Format

brettimus committed:
Modify docs tab title styles

brettimus committed:
Implement simons more recent docs designs

brettimus committed:
Update lilo worker snippets

brettimus committed:
Remove cl

brettimus committed:
Default to window.location.origin if no servers present

brettimus committed:
Print warning message if there is an unretryable error in the DOM

brettimus committed:
Add some terrible code for clearer error messages to end user when spec fails to parse

stephlow committed:
Merge remote-tracking branch 'origin/playground' into embedded-playground-api

stephlow committed:
Merge remote-tracking branch 'origin/playground' into embedded-playground-api

stephlow committed:
Specify zod dep

stephlow committed:
Remove playground dist from scm

brettimus committed:
Refactor DOM parsing stuff

brettimus committed:
Improve command bar command matching

stephlow committed:
Fix merge conflict feedback

brettimus committed:
Merge branch 'embedded-playground-api' into playground

brettimus committed:
Refactor routeDocumentation rendering and improve handling of e.g., arrays of strings

brettimus committed:
Move zod from peer deps to deps for embedded package

brettimus committed:
Improve handling of objects and arrays of objects

brettimus committed:
Run formatter and add enums to url parameter docs

brettimus committed:
Update embedded to 0.0.16

brettimus committed:
Oops update the zod dep for embedded again bc i did a dumb

brettimus committed:
Update lilo-worker to 0.0.16 of embedded

brettimus committed:
Change styles on unknown response body

brettimus committed:
Fix up the menu items along the bottom

brettimus committed:
Prepare for 0.0.17 embedded

brettimus committed:
Update lilo to latest embedded

brettimus committed:
Merge branch 'main' into playground

brettimus committed:
Fix panel overflow on small screens

brettimus committed:
Add shortcut to open keyboard shortcut modal

brettimus committed:
More radius please

brettimus committed:
No mono font for fallback title http method in route docs

brettimus committed:
Empty state for headers tab in response panel && format

brettimus committed:
Fix issue with rendering nested objects in the schema viewer for repsonse data

brettimus committed:
Update embedded version

brettimus committed:
Docs font color improvements, more keyboard shortcuts, better command menu doc page matching

brettimus committed:
Improve keyboard shortcuts modal

flenter committed:
Adding bearer token auth flow (#451)

* wip

* Add auth section

* Improve Auth stuff

* Bearer form tweak

* Minor tweak

* Refactor AuthSelector

brettimus committed:
Improve faker tooltip

brettimus committed:
Remove text-gray styles from tabs

brettimus committed:
Make our mock tightknit spec point to the mock tightknit api server

brettimus committed:
Remove double separator

brettimus committed:
Change auth section styles

brettimus committed:
Fix overflow that happened for large inputs on auth input

brettimus committed:
Add helper text about persisting to local storage

brettimus committed:
Fix type issues and fiddle with the ui for auth tokens

brettimus committed:
Fix add button

brettimus committed:
Make auth tab more kosher (with visible tabs types, etc) and add keyboard shortcut

brettimus committed:
Update embedded package

brettimus committed:
Update gitignore

brettimus committed:
Update lilo worker embedded package.json

brettimus committed:
Update npmignore for embedded package

brettimus committed:
Change playground html title and remove console log

brettimus committed:
Remove eslint from playground

brettimus committed:
Update README

brettimus committed:
Add keyboard shortcut to modal for opening the help menu

brettimus committed:
Update note in playground vite config

brettimus committed:
Add a note in the tailwind config about the custom colors we use in playground

brettimus committed:
Remove tester route

brettimus committed:
Use barrel file for RouteDocumentation

brettimus committed:
Remove unused assets, remove TODO, format

brettimus committed:
Make keyboard shortcuts modal cross-platform

brettimus committed:
Remove Branding component (unused)

brettimus committed:
Update the resizable component in playground to have optimizations from studio

brettimus committed:
Allow keybindings on codemirror inputs

laulauland committed:
WIP: Workflows in Playground 🧪 (#453)

* initial workflow setup

* setup a simple api for now

* WIP

* WIP

* remove workflows-api

* stub workeroni

* more UI updates

* lockfile

* update the workflows

* attempt to load the openapi spec like the root app

* vite redirects for dev

* remove dead schema hooks

* update data loading workflow routes

* remove workeroni

* mount the workflow routes on embedded

* rework the interface for embedded playground mounting - serialize all options into data-options attribute on <div root>

* update the hooks to load the updated data-options thing faster

* simplify types

* clean up workflow root

* load openapi in to workflow/create

* load initial three panel ui updates

* resolved routeTree.gen.ts

* update schemas so they parse correctly

* change input types

* update the detail view so it shows inputs

* resolve the input runtime expressions to values from inputs

* formattingggggg

* add command bar and load the data in the component so it re-renders (as opposed to from route)

* add a base prefix to fetch calls

* for some reason getWorkflows got yeet

* tsc complaints

* bump version

* more ts complaints

* Playground auth tweak (#454)

* Add spinner

* Implementation of early design

* Cleanup

* ship fixes from late friday

* resolve version dynamically from package.json

* update layout with top bar

* unify sidebar and use new screen without button

* add floaty workflow input - WorkflowCommand

* set default preloads for snippier effect

* formatting

---------

Co-authored-by: Jacco Flenter <flenter@users.noreply.github.com>

brettimus committed:
Update lockfile

brettimus committed:
Add cors to api to support cross-origin requests for trace data

brettimus committed:
Claude take the wheel ((let agent create the traces pages))

brettimus committed:
Port the timeline to the playground

brettimus committed:
Remove old claude ui

brettimus committed:
SAY MY NAME. I AM TIMELINE.

brettimus committed:
Move response body viewer to its own component to share with timeline

brettimus committed:
Fix nord theme

brettimus committed:
Fix weird spacing issues in the timeline with h-7

brettimus committed:
Remove weird gap

brettimus committed:
Fix sql statement viewer styles

brettimus committed:
Fix styles for fetch spans

brettimus committed:
Prep api for release with cors

brettimus committed:
Update all the zods to try to fix a type error with zodResolver in studio

brettimus committed:
Update embedded version 0.0.26

brettimus committed:
Make playground compile

brettimus committed:
Add error boundaries for fetching routes and traces

brettimus committed:
Replicate initial panel sizing from jaccos latest studio pr

brettimus committed:
Replace the fiberplane/hooks version of useMediaQuery

brettimus committed:
Replace studio implementation of useMedia to fix layout bugs on sm screens

flenter committed:
Playground routing (#458)

* Tweak routing logic

* Fix routing

* Format code

* Simplify

* Move settings to url

And remove it from the uiSlice (since it's not being used there anyway)

* Minor settings improvements

* moved SettingsScreen component to avoid an issue with aria-bidden
* tweaked behaviour so we don't end up with `settings=false` in the
  querystring

* Undo change to index.html

brettimus committed:
Merge branch 'playground' into playground-launchpad-demo

laulauland committed:
kind of working step execution

brettimus committed:
Try removing pragmas that specify hono jsx types in the playground.tsx file (resolving type error emitted by wrangler in dev)

brettimus committed:
Integrate fpx endpoint into embedded + playground

laulauland committed:
fix the useSearch bork and load openapi data more properly

brettimus committed:
Add hacky error boundaries and format

brettimus committed:
Update embedded package.json

brettimus committed:
Appease typescript

brettimus committed:
Tweaks to fetching traces (needs more work)

brettimus committed:
CORS got me down

brettimus committed:
Update embedded version

brettimus committed:
Embedded alpha

brettimus committed:
Fix requests to fp tracing

brettimus committed:
Try fetching the traces through embedded again

brettimus committed:
Fix parscing of openapi.content

laulauland committed:
fix openapi content for .new route

flenter committed:
Avoid too aggressive loading/fetching

brettimus committed:
Update crufty lilo comment

brettimus committed:
Update embedded to a standard version

laulauland committed:
fix WorkflowCommand loading openapi data

brettimus committed:
Remove cl

brettimus committed:
Convert Trace available callout to an error banner

laulauland committed:
add /report route prox

brettimus committed:
Add error reporting prototype

brettimus committed:
Format and actually register the report creation api

brettimus committed:
Update playgroudn vite config to be able to configure the proxy url for the embedded api

brettimus committed:
Add a success message when issue reported

flenter committed:
Use @apidevtools/swagger-parser

* Add typescript configuration
* Add typescript checks to `lint` command
* Improve the output value/parameter logic
* add details to ErrorBoundary

brettimus committed:
Spruce up the request sharing form

brettimus committed:
Prepare embedded 0.0.29-alpha.2

flenter committed:
Merge remote-tracking branch 'origin/playground-launchpad-demo' into playground-launchpad-demo-workflow-execute

laulauland committed:
TEMP: enable hiding the traces navbar

flenter committed:
Cleanup

flenter committed:
Merge remote-tracking branch 'origin/playground-launchpad-demo' into playground-launchpad-demo-workflow-execute

brettimus committed:
Fix overflow of response summary in the tab

brettimus committed:
Fix some styles on error logs in the timeline

brettimus committed:
Change the log even indicator to an orange circle

brettimus committed:
Remove route list from trace details page

brettimus committed:
Connect the trace  summary stuff

flenter committed:
Playground launchpad demo: workflow execute (#461)

Co-authored-by: Laurynas Keturakis <laurynas.keturakis@gmail.com>

flenter committed:
Initial UI tweak

brettimus committed:
Remove unused

brettimus committed:
Fix embedded build and tweak gitignore to ignore compiled viteconfig

brettimus committed:
Update embedded and playground for well rendered ai summaries on the right

flenter committed:
Add minimal controls to steps list

flenter committed:
Fix lint issues

brettimus committed:
Revert jacco pr real quick

brettimus committed:
Bump embedded to release

brettimus committed:
New alpha

flenter committed:
Switch back to @scalar/openapi-parser

flenter committed:
Fix runtime issue

flenter committed:
Update hook

flenter committed:
Fix build process

flenter committed:
Fix tsconfig

flenter committed:
Tweak styling a bit more

flenter committed:
Further UI tweaks

flenter committed:
Merge remote-tracking branch 'origin/playground-launchpad-demo' into playground-launchpad-demo-workflow-execute-ui

flenter committed:
Bump package version of embedded

brettimus committed:
Do not send body with GET or HEAD requests

brettimus committed:
Alpha 8 beebee

flenter committed:
Improve the ui a bit further

flenter committed:
Add shake animation

flenter committed:
Use danger

flenter committed:
Tweak UI further

flenter committed:
Remove commented out code

brettimus committed:
Merge branch 'main' into playground

brettimus committed:
Move garbage/RequestorPage into "components/playground"

brettimus committed:
Update README.md for playground

Squash

brettimus committed:
Hide the mock api spec switch in prod

brettimus committed:
Add back SettingsScreen

brettimus committed:
Remove Use Mock Spec from CommandBar and fix settings page

brettimus committed:
Feature-flag the workflows and tracing, conditionally hide the top nav

brettimus committed:
Remove unused imports and unused utils.ts file

brettimus committed:
Rename "useRequestorSubmitHandler" to "usePlaygroundSubmitHandler"

brettimus committed:
Rename RequestorPage to PlaygroundPage and remove dead styles.css

brettimus committed:
Rename RequestorInput to PlaygroundInput

brettimus committed:
Rename prop "isRequestorRequesting" to "isPlaygroundRequesting"

brettimus committed:
Rename playground/store schemas and types relating to Requestor* to Playground*

brettimus committed:
Remove all remaining references to requestor

brettimus committed:
Put error reporting behind a feature flag

brettimus committed:
Remove references to ProxyRequestResponse

brettimus committed:
Handle request failures

brettimus committed:
Rename useMakeProxiedRequest to useMakePlaygroundRequest

brettimus committed:
Fix issue where the path params were not updating when you used fillInFakeData

brettimus committed:
Turn off tracing in redux dev tools

brettimus committed:
Playground: Routes sidebar improvements (#469)

* Rename ProbedRoute to ApiRoute

* Remove isDraft from ApiRoute type

* Remove the "routeOrigin" field from ApiRoute

* Remove concept of route vs middleware, currentlyRegistered, and handler code from ApiRoute type

* Format

* Rip out requestType from ApiRoute and helper functions (which could have been websocket instead of http)

* Remove type that was geared towards WS requests and add support for TRACE http method

* Remove concept of registrationOrder from ApiRoute type

* Add summary, description, and tags to the ApiRoute

* Remove concept of routesAndMiddleware from the hooks and store

* First pass of rendering route tags / summaries in the RoutesPanel

* Render "DEL" instead of "DELETE" in the RoutesPanel for a route item

* When all routes are untagged, just render the list of routes

* Use a symbol to represent untagged routes

flenter committed:
Use CustomTabs as the standard tabs (#470)

* Make CustomTabs the default tabs

And rename them to `FpTabs...`

* Set display name again

brettimus committed:
Rip out use of Hono router in the UI for path matching logic (#471)

* Change specToApiRoutes to preserve openapi-style paths

* Add basic support in the UI for OpenAPI-style path params

* Remove Hono-router-related route matching logic

* Fix insertion of fake-data path params

* Add playground tests to CI

brettimus committed:
Make the settings page internally scrollable

brettimus committed:
Change the copy for experimental features to mention contacting fiberplane

brettimus committed:
Bump embedded package version to 0.0.30-alpha.1

brettimus committed:
Add back DialogDescription for screen readers on SettingsScreen

brettimus committed:
Respect ordering of OpenAPI tags in the RoutesPanel

laulauland committed:
Playground: adapt to tenancy related work (#472)

* pluralization

* add alias

* lift the URL to consts

* rename to plural

* use factory pattern so we can pass in the api key

* rename all the things

* rename again

* rename for consistency

* remove for possible errors

* workflowZZZZZ

* use Tanstack API to ensure typechecking on custom link comps

* fixes for the fix god

laulauland committed:
package.json

flenter committed:
Fix missing broken workflow links

laulauland commented on playground/src/lib/tightknit.json:
just a note to remove this before merging

brettimus committed:
Remove comments

brettimus committed:
Respect the `content` prop, gate `/fp/api`, modularize the playground api client, and more! (#473)

* Respect the "content" property of openapi options from middleware when parsing a spec in the DOM

* Make apiKey optional and document it with typescript doc comment

* Remove crufty console.logs from tracing api

* Bump embedded package.json to alpha.6

* Only mount api routes if apiKey is present

* Fix package.json

* Tweak docstrings

* Modify NavButtonComponent in Layout to use a forwardRef and run biome format

* Change <ThemeProvider>s to use system as default theme

* Add a specific error screen for if Workflows are not enabled

* Improve the error screen when the OpenAPI spec cannot be fetched

* Factor out lib/api into its own module

* Try using new fetch wrappers in api client

* Start pulling out utils that might be causing a circular dependency error

* Resolve circular dependencies

* Use the new fetch wrappers

* Clean up ErrorScreen and FeatureDisabledScreen props

* Remove references to fpxEndpointHost

brettimus committed:
Remove console.log

brettimus committed:
Bump embedded alpha

brettimus committed:
Bump package.json to 0.0.30 for embedded package

brettimus committed:
Made theme selection component better

brettimus committed:
Remove unused package from embedded package.json

flenter committed:
Add scratchpad functionality (#475)

* Initial implementation of scratchpad

* Avoid several re-renders

* Move response to the ApiCallData object

* Show preview of url

avoid resetting when url is updated with params

* Fix actual requesting with parameters

And also disable editing on url input

* Remove method selector interactivity

* Turn the PlaygroundInput into a dropdown

* Remove updatePath and updateMethod from store

* Halfway through removing path/method from store

* Second part: remove path/method from store

* Add useCurrentAuthorization hook

And remove quite a few lines of commented out code

* Fix formatting

* Cleanup

* Add docstrings explaining the apiCallState

* Remove "pt-0.5" from Method component, and render DEL instead of DELETE

* Add a helper for shortening the HTTP method

* Use helper for shortening HTTP method for ui

* Make routes dropdown less padded

* Fix truncation on the PlaygroundInput

* Add a min width to the PlaygroundInput Method to avoid horizontal jumping

* Remove comment out code

* Remove commented out code and realign a comment

* Remove session history from RequestResponseSlice

* Remove old commented out throw

* Remove unnecessary conditional

* Rename

---------

Co-authored-by: Brett Beutell <brbeut@gmail.com>

mellowagain committed:
Fp 4331 add debug logging to the embedded middleware (#478)

* add debug logging

* return correct type

* Format

* Spread the params back into the console.debug statement

* Modify logIfDebug to accept Context param as well

* Support addtl params in logIfDebug, rename debugEnabled to isDebugEnabled, and modify a few debug logs

* Modify some of the routing debug logs

* Add vitest to embedded

* Move logIfDebug to its own file

* Write a test file for logIfDebug

* Add docstring to logIfDebug

---------

Co-authored-by: Brett Beutell <brbeut@gmail.com>

brettimus committed:
Remove option to use a mock api spec in the playground (#476)

* Made theme selection component better

* Remove option to use a mock api spec in playground ui

brettimus committed:
Prepare embedded package for release (#479)

* Rename createMiddleware to createFiberplane and try to use hono factor to resolve downstream type errors

* Simplify middleware factory to fix typing issue with published package

* Fall back to FIBERPLANE_API_KEY if present, and document this behavior in the types

* Add some more debug logs to the middleware

* Bump package.json

* Still export createMiddleware so we do not break the readmes we gave to people

* Rename embedded package name to @fiberplane/hono and change version to 0.4.0

* Update CDN url to @fiberplane/hono for embedded assets

* Update all crufty references to @fiberplane/embedded

* Update lilo to use latest @fiberplane/hono

* Remove @fiberplane/hono from the studio api

brettimus committed:
Add links to GH, Discord, and tweak default button styles (#480)

* Modify default button styles

* Clean up styles after button default change

* Hide user Icon in the top nav

* Add GitHub and Fiberplane links to the bottom bar

* First pass at adding branding to bottom nav

* Add link to Discord in the settings screen to request api key

* Fix typo

* Remove unnecessary ts-expect-error

* Improve discord icon and external link tooltips

* Update dev favicon

* Update embedded package.json to a beta version

* Fix weird layout issue with input in production ((regression))

* Bump embedded package.json again

* Use nice theme illustrations from Simon on settings page

* Remove border from theme selection and rename the component

* Update embedded to beta.3

* Update package.json to 0.4.1

brettimus committed:
Merge branch 'main' into playground

brettimus committed:
Merge branch 'main' into playground

brettimus committed:
Reorg the monorepo (#483)

* Remove scripts/otel_collector

* Move python code into a python/ folder

* Move webhonc, honc-code-gen, fp-services into apps/ dir and update biome,pnpm accordingly

* Fix type issue with honc code gen

* Move studio frontend and api to own folder, remove lilo, and update pnpm,package.json,biome.jsonc

* Update github workflows

* Fix studio README

* Rename embedded, add symbolic link to fiberplane-hono package README

* Copy old DEVELOPMENT file into apps/studio

* Remove .fpx folder

* Resolve type errors in studio-api and webhonc

* Update all the drizzles

* Move playground into packages/ and update package.json files that pointed to the old fpx url

* Rename "embedded" script

* Remove local-inference

* Try linking to a new logo

* Put branding in README

* Write simple readme

* Update DEVELOPMENT.md

* Modify the DEVELOPMENT.md

* Update README.md

* Update README.md

* Update README.md

---------

Co-authored-by: Micha Hernandez van Leuffen <m.hernandezvanleuffen@gmail.com>

flenter committed:
Fix biome configuration paths

flenter reviewed (approved):
I'm glad some of the code has been reviewed before merging into this pr, but this PR is too big to be reviewed in detail. I will approve it so we can move forward.
